Bump Sentry Gradle plugin

wzieba opened the pull request:
## Description

This PR updates Sentry Gradle plugin to newest stable version, adding new features and improving build time.

### Source context upload

Different issues, but you can see that new issues will show the source code as well.

| Before | ![image](https://github.com/wordpress-mobile/WordPress-Android/assets/5845095/4f38cad0-3abd-449a-a22f-7ae4a82467eb) |
| --- | --- | 
| After | ![image](https://github.com/wordpress-mobile/WordPress-Android/assets/5845095/7924b22a-857e-4cda-b5c9-f7dfbe4236fa)|

### Disable dependencies report

We can always check the release and see what dependencies were used in a specific build that Sentry issue is coming from. This saves us from additional tasks in the build tree and makes the Sentry payload for each issue smaller.

### Disable Sentry tasks in `debug`

We don't use Sentry in debug builds. This removes all Sentry tasks in this build type, making builds quicker.

## To Test:

Please build the release variant (`vanillaRelease`) of both WordPress and Jetpack flavors and verify if issues are sent correctly. You can use the following snippet.

<details><summary>PATCH</summary>

```PATCH
Subject: [PATCH] tests: add a test for features in development generation
---
Index: WordPress/src/main/java/org/wordpress/android/viewmodel/main/WPMainActivityViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WordPress/src/main/java/org/wordpress/android/viewmodel/main/WPMainActivityViewModel.kt b/WordPress/src/main/java/org/wordpress/android/viewmodel/main/WPMainActivityViewModel.kt
--- a/WordPress/src/main/java/org/wordpress/android/viewmodel/main/WPMainActivityViewModel.kt	(revision 806913d9fb807250cecd5b24b36001d55ea4c255)
+++ b/WordPress/src/main/java/org/wordpress/android/viewmodel/main/WPMainActivityViewModel.kt	(date 1710772966823)
@@ -5,6 +5,7 @@
 import androidx.lifecycle.LiveData
 import androidx.lifecycle.MutableLiveData
 import androidx.lifecycle.distinctUntilChanged
+import com.automattic.android.tracks.crashlogging.CrashLogging
 import kotlinx.coroutines.CoroutineDispatcher
 import kotlinx.coroutines.flow.firstOrNull
 import org.wordpress.android.R
@@ -67,6 +68,7 @@
     private val bloggingPromptsStore: BloggingPromptsStore,
     @Named(UI_THREAD) private val mainDispatcher: CoroutineDispatcher,
     private val shouldAskPrivacyConsent: ShouldAskPrivacyConsent,
+    private val crashLogging: CrashLogging,
 ) : ScopedViewModel(mainDispatcher) {
     private var isStarted = false
 
@@ -161,6 +163,7 @@
         launch { loadMainActions(site) }
 
         updateFeatureAnnouncements()
+        crashLogging.sendReport(Throwable("Test crash"))
     }
 
     @Suppress("LongMethod")
```
</details>

-----

## Regression Notes

1. Potential unintended areas of impact

    - Sending Sentry issues

2. What I did to test those areas of impact (or what existing automated tests I relied on)

    - Manual tests

3. What automated tests I added (or what prevented me from doing so)

    - None

-----

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## Testing Checklist (strike-out the not-applying and unnecessary ones):

- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

wzieba committed:
build: bump Sentry Gradle Plugin to 4.3.1

wzieba committed:
style: improve formatting of Sentry configuration

wzieba committed:
build: upload source context

This will send source code to Sentry, so we can see the code that caused the issue directly on Sentry dashboard

wzieba committed:
build: do not generate list of dependencies for Sentry

We know the list of dependencies as we attach information about release to each ticked.

Disabling this feature improves build times.

wzieba committed:
build: do not run any Sentry tasks in debug builds

To save local build time

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba committed:
Merge branch 'trunk' into bump_sentry_gradle_plugin

wzieba committed:
fix: when copying React Native bundle source map, rely on input/output of copy task

It worked previously, as Sentry Gradle plugin was making the merge<variant>Assets run every time (breaking up to date state)

wzieba committed:
Merge branch 'trunk' into bump_sentry_gradle_plugin

irfano reviewed (approved):
Great job! I could see my "Test crash" in the Sentry report.
Thank you for adding a patch for testing. I'm approving but could you also paste "To Test" section of the PR into [Test Instructions per Dependency Update](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/docs/test_instructions_per_dependency_update.md#sentry-sentryversion-) page before merging the PR.

wzieba commented:
> could you also paste "To Test" section of the PR into [Test Instructions per Dependency Update](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/docs/test_instructions_per_dependency_update.md#sentry-sentryversion-) page before merging the PR.

Sure, I'll do!

wzieba committed:
docs: add 'how to test Sentry' section to Test Instructions per Dependency Update

wzieba commented on docs/test_instructions_per_dependency_update.md:
````diff
@@ -854,9 +854,58 @@ and [here](https://github.com/orgs/wordpress-mobile/projects/95)).
+
+<details><summary>PATCH (warning: it'll probably have some conflicts in the future when `WPMainActivityViewModel` change. It's more for an idea:</summary>
+
+```PATCH
+Subject: [PATCH] tests: add a test for features in development generation
+---
+Index: WordPress/src/main/java/org/wordpress/android/viewmodel/main/WPMainActivityViewModel.kt
+IDEA additional info:
+Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
+<+>UTF-8
+===================================================================
+diff --git a/WordPress/src/main/java/org/wordpress/android/viewmodel/main/WPMainActivityViewModel.kt b/WordPress/src/main/java/org/wordpress/android/viewmodel/main/WPMainActivityViewModel.kt
+--- a/WordPress/src/main/java/org/wordpress/android/viewmodel/main/WPMainActivityViewModel.kt	(revision 806913d9fb807250cecd5b24b36001d55ea4c255)
++++ b/WordPress/src/main/java/org/wordpress/android/viewmodel/main/WPMainActivityViewModel.kt	(date 1710772966823)
+@@ -5,6 +5,7 @@
+ import androidx.lifecycle.LiveData
+ import androidx.lifecycle.MutableLiveData
+ import androidx.lifecycle.distinctUntilChanged
++import com.automattic.android.tracks.crashlogging.CrashLogging
+ import kotlinx.coroutines.CoroutineDispatcher
+ import kotlinx.coroutines.flow.firstOrNull
+ import org.wordpress.android.R
+@@ -67,6 +68,7 @@
+     private val bloggingPromptsStore: BloggingPromptsStore,
+     @Named(UI_THREAD) private val mainDispatcher: CoroutineDispatcher,
+     private val shouldAskPrivacyConsent: ShouldAskPrivacyConsent,
++    private val crashLogging: CrashLogging,
+ ) : ScopedViewModel(mainDispatcher) {
+     private var isStarted = false
+
+@@ -161,6 +163,7 @@
+         launch { loadMainActions(site) }
+
+         updateFeatureAnnouncements()
++        crashLogging.sendReport(Throwable("Test crash"))
+     }
+
+     @Suppress("LongMethod")
+```
+</details>
````
WDYT @irfano ?

irfano replied:
It looks good! Thank you for including additional comments. 👍🏻
